### PR TITLE
public comments on validation rules

### PR DESCRIPTION
**Pull request by WFinck97**

### CODE REVIEW

**Review by Connoropolous — CHANGES_REQUESTED** [review]
can you make the comment more meaningful?

File: dna/zomes/projects/src/project/project_meta/validate.rs [hunk 1]
```diff
@@ -25,6 +26,7 @@ pub fn validate_create_entry_project_meta(
 }
 
 #[hdk_extern]
+/// Updates allowed by anyone, but only `name` and `image` can be updated
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/project_meta/validate.rs:29]: "but only `name` and `image` can be updated"

that used to be the case, but I reversed in a way to validate ones that can't change, as opposed to ones that can... so 
if project_meta.creator_address == original_project_meta.creator_address
                        && project_meta.created_at == original_project_meta.created_at
                        && project_meta.passphrase == original_project_meta.passphrase
creator_address, created_at, and passphrase can't change, the others can.

File: dna/zomes/projects/src/project/project_meta/validate.rs [hunk 2]
```diff
@@ -6,6 +6,7 @@ use crate::project::{
 use hdk::prelude::*;
 
 #[hdk_extern]
+/// Creates allowed if `creator_address` matches the agent committing
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/project_meta/validate.rs:9]: 👍

File: dna/zomes/projects/src/project/member/validate.rs [hunk 3]
```diff
@@ -4,6 +4,7 @@ use crate::project::{
 use hdk::prelude::*;
 
 #[hdk_extern]
+/// Creates allowed if the member `address` matches the author of the entry
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/member/validate.rs:7]: 👍 
this data type declares themselves to be a member of the project. we do a quick and easy lookup of these off an anchor to check who's in

File: dna/zomes/projects/src/project/goal_vote/validate.rs [hunk 4]
```diff
@@ -35,6 +37,8 @@ pub fn validate_create_entry_goal_vote(
 }
 
 #[hdk_extern]
+/// Updates allowed if Goal exists and only by the original author
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/goal_vote/validate.rs:40]: `must_get_header_and_entry::<GoalVote>`

so it's the original GoalVote that must exist not the Goal... we assume the Goal existed if the validation for create had passed...

File: dna/zomes/projects/src/project/goal_vote/validate.rs [hunk 5]
```diff
@@ -9,6 +9,8 @@ use crate::project::{
 use hdk::prelude::*;
 
 #[hdk_extern]
+/// Votes can only be created if the Goal exists and if `agent_address` of the
+/// Vote matches the agent authoring the enrty, unless the vote is imported
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/goal_vote/validate.rs:13]:
```suggestion
/// `GoalVote`s can only be created if the Goal exists and if `agent_address` of the
/// `GoalVote` matches the agent authoring the entry, unless the `GoalVote` is imported
```

File: dna/zomes/projects/src/project/goal_comment/validate.rs [hunk 6]
```diff
@@ -35,6 +38,8 @@ pub fn validate_create_entry_goal_comment(
 }
 
 #[hdk_extern]
+/// Comments can only be updated by the original commentor and the Goal must exists.
+/// `agent_address` should not change from the original value
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/goal_comment/validate.rs:42]:
```suggestion
/// `GoalComment`s can only be updated by the original commenter and the Goal must exist.
/// `agent_address` should not change from the original value
```

File: dna/zomes/projects/src/project/goal_comment/validate.rs [hunk 7]
```diff
@@ -9,6 +9,9 @@ use crate::project::{
 use hdk::prelude::*;
 
 #[hdk_extern]
+/// Goal comments can only be created if the Goal exists and
+/// the `agent_address` must match the address of the agent adding
+/// the comment, unless the entry is imported
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/goal_comment/validate.rs:14]:
```suggestion
/// `GoalComment`s can only be created if the `Goal` exists and
/// the `agent_address` must match the address of the agent adding
/// the `GoalComment`, unless the entry is imported (`is_imported`)
```

File: dna/zomes/projects/src/project/goal/validate.rs [hunk 8]
```diff
@@ -43,6 +45,8 @@ pub fn validate_create_entry_goal(
 }
 
 #[hdk_extern]
+/// Updates only allowed if `user_edit_hash` of the updated Goal entry matches the
+/// agent publishing the update
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/goal/validate.rs:49]: Can you also add that `user_hash` can't be modified (the address of the "original author" can't change)

File: dna/zomes/projects/src/project/goal/validate.rs [hunk 9]
```diff
@@ -11,6 +11,8 @@ use hdk::prelude::*;
 use hdk_crud::WrappedAgentPubKey;
 
 #[hdk_extern]
+/// Creates only allowed if `user_hash` of the Goal entry matches the agent
+/// creating the entry, unless the Goal is imported
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/goal/validate.rs:15]: Can you also add that: `user_edit_hash` must not be set (before "unless the Goal is imported")

File: dna/zomes/projects/src/project/entry_point/validate.rs [hunk 10]
```diff
@@ -5,6 +5,9 @@ use crate::project::{error::Error};
 use hdk::prelude::*;
 
 #[hdk_extern]
+/// A goal must exist at the entry point and the creator
+/// of the entry point must be the author of the goal unless
+/// the entry is imported
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/entry_point/validate.rs:10]: Almost. The `creator_address` of the EntryPoint should match the agent authoring the EntryPoint. 

And you could say the `Goal` at the `goal_address` on the `EntryPoint` must exist, as opposed to "A goal must exist at the entry point"

File: dna/zomes/projects/src/project/edge/validate.rs [hunk 11]
```diff
@@ -3,6 +3,7 @@ use crate::project::error::Error;
 use hdk::prelude::*;
 
 #[hdk_extern]
+/// New edges cannot be self-referential and both a parent and child goal must exist
```
Comment (Connoropolous) [inline dna/zomes/projects/src/project/edge/validate.rs:6]:
```suggestion
/// New edges cannot be self-referential and both the parent and child goal must exist
```

**Review by Connoropolous — CHANGES_REQUESTED** [review]
Good start! 

It can be typical to have something like this where there's lots of feedback on PRs, in the early stage of collaborating. Think of PRs and reviews as a form of communication, back and forth, between developers, focused on the artifact of the code. 

Some of my comments were easy to express directly as edits, in which case, you can just 'merge them' directly within Github. If I didn't make a "suggested change" in the form of code, then it'll be up to you to put something in. Each comment should be resolved through either directly merging them, or sparking further conversation in the PR by making comments and questions. Eventually each point should get resolved.

Give it a try

File: dna/zomes/projects/src/project/project_meta/validate.rs (hunk 1 above)
Comment (WFinck97) [reply dna/zomes/projects/src/project/project_meta/validate.rs:29]:
```suggestion
/// Updates allowed by anyone, but only `name` and `image` can be updated whereas `creator_address`, `created_at`, and `passphrase` can't change
```

Comment (Connoropolous) [reply dna/zomes/projects/src/project/project_meta/validate.rs:29]: There's a couple more fields that can change, that have been added. Just make the comment discuss the ones that can't be changed

File: dna/zomes/projects/src/project/entry_point/validate.rs (hunk 10 above)
Comment (WFinck97) [reply dna/zomes/projects/src/project/entry_point/validate.rs:10]:
```suggestion
/// The `Goal` at the `goal_address` on the `EntryPoint` must exist and the `creator_address` of the
/// `EntryPoint` must match the agent authoring the `EntryPoint`, unless the entry is imported
```

File: dna/zomes/projects/src/project/goal/validate.rs (hunk 9 above)
Comment (WFinck97) [reply dna/zomes/projects/src/project/goal/validate.rs:15]:
```suggestion
/// creating the entry and if `user_edit_hash` is not set, unless the Goal is imported 
```

File: dna/zomes/projects/src/project/goal/validate.rs (hunk 8 above)
Comment (WFinck97) [reply dna/zomes/projects/src/project/goal/validate.rs:49]:
```suggestion
/// agent publishing the update, `user_hash` cannot change from the original value
```

**Comment by Connoropolous** [comment]
Yep @WFinck97 just commit all these using the 'add to batch' thing, and then I'll re-review, though I think everything's good. It'll just be an approve at that point

File: dna/zomes/projects/src/project/goal_vote/validate.rs (hunk 4 above)
Comment (WFinck97) [reply dna/zomes/projects/src/project/goal_vote/validate.rs:40]:
```suggestion
/// Updates allowed if the original `GoalVote` exists and only by the original author
```

File: dna/zomes/projects/src/project/project_meta/validate.rs (hunk 1 above)
Comment (WFinck97) [reply dna/zomes/projects/src/project/project_meta/validate.rs:29]:
```suggestion
/// Updates allowed by anyone, but `creator_address`, `created_at`, and `passphrase` cannot be changed
```

**Review by Connoropolous — APPROVED** [review]
Sweet! First PR

**Pull request merged**